trinity/sync/beacon/chain.py

```diff
@@ -0,0 +1,145 @@
+from typing import (
+    cast,
+    Tuple,
+    AsyncGenerator,
+)
+
+from eth_utils import (
+    ValidationError,
+)
+from eth_utils.toolz import (
+    first,
+)
+
+from cancel_token import (
+    CancelToken,
+)
+
+from p2p.service import (
+    BaseService,
+)
+
+from eth.beacon.types.blocks import BaseBeaconBlock
+from eth.beacon.db.chain import BaseBeaconChainDB
+
+from trinity.protocol.bcc.peer import (
+    BCCPeer,
+    BCCPeerPool,
+)
+from trinity.sync.beacon.constants import (
+    MAX_BLOCKS_PER_REQUEST,
+)
+
+
+class BeaconChainSyncer(BaseService):
+    """Sync from our finalized head until their preliminary head."""
+
+    def __init__(self,
+                 chain_db: BaseBeaconChainDB,
+                 peer_pool: BCCPeerPool,
+                 token: CancelToken = None) -> None:
+        super().__init__(token)
+
+        self.chain_db = chain_db
+        self.peer_pool = peer_pool
+
+        self.sync_peer: BCCPeer = None
+
+    async def _run(self) -> None:
+        self.sync_peer = self.select_sync_peer()
+        if not self.sync_peer:
+            self.logger.info("No peers to sync with")
+            return
+
+        if not self.sync_peer_useful():
+            self.logger.info("Not starting sync as no peer is useful")
+            return
+
+        await self.sync()
+
+        new_head = self.chain_db.get_canonical_head()
+        self.logger.info(f"Sync with {self.sync_peer} finished, new head: {new_head}")
+
+    def select_sync_peer(self) -> BCCPeer:
+        if len(self.peer_pool) == 0:
+            raise ValueError("No peers to sync with")
+        peers = [
```

Comment:
`sorted` works fine with generators.  So you can convert this to be something like `(peer for peer in self.peer_pool.connected_nodes.values())`

Also, it might be a little cleaner to do `cast(Iterable[BCCPeer], peers)` assuming that works well enough with `mypy`.